Extend abi check functionality

dgreen-arm opened the pull request:
## Description
This aims to extend the ABI checking script in a number of ways to more easily check we don't break the ABI with the Crypto submodule.

- Add ability to compare between different repositories.
- Add skipping internal identifiers.
- Add handling for the Crypto submodule.

## Status
**READY**

Patater commented on scripts/abi_check.py:
```diff
@@ -231,17 +256,37 @@ def run_main():
+            old_rev = abi_args.old_rev[0]
+        elif len(abi_args.old_rev) == 2:
+            old_repo = abi_args.old_rev[0]
+            old_rev = abi_args.old_rev[1]
```
Would it be easier to understand and more readable to have "--old-rev" and "--old-repo" options instead of using repeated "--old-rev" to specify the repo? At least users wouldn't have to remember which order to use.

dgreen-arm replied:
You're probably right. There's a lot of options already, a few more won't hurt.

Patater commented on scripts/abi_check.py:
```diff
@@ -114,11 +118,24 @@ def update_git_submodules(self, git_worktree_path):
         my_environment = os.environ.copy()
         my_environment["CFLAGS"] = "-g -Og"
         my_environment["SHARED"] = "1"
+        my_environment["USE_CRYPTO_SUBMODULE"] = "1"
```
Is this going to always build with the submodule for versions of TLS that supporting building with submodule? This is a change worth mentioning in the commit message.

dgreen-arm replied:
As it is, it will always try to use the submodule. Do you think its worth adding options to enable testing with and without the submodule?

Patater replied:
Always using the submodule makes sense, since we want to test Crypto mainly (and TLS will only be able to use the submodule after `mbedtls_*` crypto functions move there) as does falling back to not use the submodule if it isn't available.

Patater replied:
Please add a note in the commit message that introduces this change.

dgreen-arm replied:
Updated the commit message.

Patater commented on scripts/abi_check.py:
```diff
@@ -31,8 +31,8 @@
     def __init__(self, report_dir, old_repo, old_rev, old_crypto_rev,
-                 new_repo, new_rev, new_crypto_rev, keep_all_reports, brief,
-                 skip_file=None):
+                 old_crypto_repo, new_repo, new_rev, new_crypto_rev,
```
Commit "Extend crypto submodule selection further to choose repo" could use a better explanation of what's going on.

Patater replied:
OK

Patater commented on scripts/abi_check.py:
```diff
@@ -120,16 +123,30 @@ def update_git_submodules(self, git_worktree_path, crypto_rev):
-            if checkout_process.returncode != 0:
-                raise Exception("git checkout failed, aborting")
+            if crypto_repo:
+                shutil.rmtree(os.path.join(git_worktree_path, "crypto"))
```
Probably don't want to do this except as part of a temporary work tree. It'd be a bummer to remove someones work in the crypto submodule with this. Do we always work with a temporary work tree?

dgreen-arm replied:
Yes, it always creates a temporary worktree. Even if it wouldn't get removed here, it would get removed in the `cleanup_worktree` function.

Patater replied:
OK

Patater commented on scripts/abi_check.py:
```diff
@@ -120,16 +123,30 @@ def update_git_submodules(self, git_worktree_path, crypto_rev):
+                    stdout=subprocess.PIPE,
+                    stderr=subprocess.STDOUT
+                )
+                clone_output, _ = clone_process.communicate()
```
It'd be much more efficient than doing a new clone if we added a new remote for the repo requested, fetched through there, and then did a checkout. `git remote add repo git@blahblah:blah/blah.git && git fetch repo && git checkout deadbeef123`. All hashes and branches already fetched wouldn't need to get fetched again; only what's different would be fetched.

dgreen-arm replied:
I'll try this and see what its like.

Patater replied:
How was it?

dgreen-arm replied:
This is much better, I've reworked it.

Patater commented on scripts/abi_check.py:
```diff
@@ -371,46 +379,12 @@ def run_main():
+            abi_args.old_crypto_rev, abi_args.old_crypto_repo,
+            abi_args.new_repo, abi_args.new_rev, abi_args.new_crypto_rev,
+            abi_args.new_crypto_repo, abi_args.keep_all_reports,
+            abi_args.brief, abi_args.skip_file
```
That is so much simpler to parse.

dgreen-arm replied:
Given that I'll need to rebase to fix commit messages, would it be worth rebasing this commit into the two commits that added the repository selections initially?

Patater replied:
That'd make reviewing commit by commit easier.

Patater replied:
This is optional to do.

dgreen-arm commented on scripts/abi_check.py:
```diff
@@ -9,10 +9,10 @@
-and 2 if there is an error while running the script.
-Note: must be run from Mbed TLS root.
+The results of the comparison are either formatted as HTML and stored at
+a configurable location, or a brief list of problems found is output.
```
I've reworded this following your suggestion.

dgreen-arm commented on scripts/abi_check.py:
```diff
@@ -23,20 +23,34 @@
-        self.should_keep_report_dir = os.path.isdir(self.report_dir)
+        self.can_remove_report_dir = not (os.path.isdir(self.report_dir) or
+                                          keep_all_reports)
+        self.old_repo = old_repo
```
I've added a new class to hold the duplicated arguments for the two versions and reworked the script accordingly.

dgreen-arm commented on scripts/abi_check.py:
```diff
@@ -58,13 +72,33 @@ def check_abi_tools_are_installed(self):
-        self.log.info(
-            "Checking out git worktree for revision {}".format(git_rev)
-        )
+    def get_clean_worktree_for_git_revision(self, remote_repo, git_rev):
```
I've made this change for the internal methods.

dgreen-arm commented on scripts/abi_check.py:
```diff
@@ -86,33 +121,67 @@ def update_git_submodules(self, git_worktree_path):
         self.log.info(output.decode("utf-8"))
         if process.returncode != 0:
             raise Exception("git submodule update failed, aborting")
+        if (os.path.exists(os.path.join(git_worktree_path, "crypto"))
```
Good catch, I've made this change.

dgreen-arm commented:
I've reworked the PR based on feedback. As some of the requested changes were for better commit messages, I've needed to rebase. There isn't any change to the content of the initial 8 commits.

Patater commented on scripts/abi_check.py:
```diff
@@ -23,20 +23,34 @@
 class AbiChecker(object):
 
-    def __init__(self, report_dir, old_rev, new_rev, keep_all_reports):
+    def __init__(self, report_dir, old_repo, old_rev, old_crypto_rev,
```
A comment referring them to the command line parameter documentation would be good. We should try to avoid duplicating documentation in this case.

Patater commented on scripts/abi_check.py:
```diff
@@ -28,7 +28,7 @@
 
     def __init__(self, report_dir, old_repo, old_rev, new_repo, new_rev,
-                 keep_all_reports):
+                 keep_all_reports, skip_file=None):
```
OK

dgreen-arm commented:
I believe I've addressed all the feedback now. It required a rebase due to changing a commit message, but there should be no code changes except for whats added in the last 2 commits.

dgreen-arm commented:
This needs backports and an issue was found when cleaning up, so removing the `ready for merge` label

dgreen-arm commented:
I've rebased on the latest development, which contains #2464. The previous branch is available here: https://github.com/dgreen-arm/mbedtls/tree/extend-abi-check-functionality-1

dgreen-arm commented on scripts/abi_check.py:
```diff
@@ -23,30 +23,37 @@
+        self._setup_logger()
+        self.report_dir = os.path.abspath(configuration.report_dir)
+        self.keep_all_reports = configuration.keep_all_reports
+        self.can_remove_report_dir = not (os.path.isdir(self.report_dir) or
```
No, that check could be written better. What I'm wanting to do there is check with `os.path.isdir` if the directory already exists. If it does already exist, then we don't want to remove it.

There is a flaw though, that at no point am I checking that it is actually a directory. I'll fix that and make it clearer.

k-stachowiak replied:
I understand the test now, thanks.

dgreen-arm replied:
I've added a check to ensure the report directory isn't a file. I also made the test a bit clearer, hopefully.

gilles-peskine-arm commented on scripts/abi_check.py:
```diff
@@ -23,30 +23,37 @@
-        old_rev: reference git revision to compare against
-        new_rev: git revision to check
         keep_all_reports: if false, delete old reports
+        brief: if true, output shorter report to stdout
```
```suggestion
        configuration.brief: if true, output shorter report to stdout
```
etc.

gilles-peskine-arm commented on scripts/abi_check.py:
```diff
@@ -68,83 +78,142 @@ def check_abi_tools_are_installed():
-        )
         git_worktree_path = tempfile.mkdtemp()
+        if version.repository:
+            self.log.debug(
```
I think this block simplifies to
```python
self.log.debug(Checking out git worktree for revision {} from {}", version.revision, version.repository)
fetch_output = subprocess.check_output(…, encoding='utf-8')
self.log.debug(fetch_output)
worktree_rev = "FETCH_HEAD"
```
Likewise for other uses of `subprocess.Popen`. In modern Python, most uses of `subprocess` boil down to `check_call` or `check_output`.

gilles-peskine-arm commented on scripts/abi_check.py:
```diff
@@ -68,83 +78,142 @@ def check_abi_tools_are_installed():
         make_output, _ = make_process.communicate()
-        self.log.info(make_output.decode("utf-8"))
+        self.log.debug(make_output.decode("utf-8"))
+        for root, dirs, files in os.walk(git_worktree_path): # pylint: disable=unused-variable
```
```suggestion
        for root, _dirs, files in os.walk(git_worktree_path):
```

gilles-peskine-arm commented on scripts/abi_check.py:
```diff
@@ -254,18 +359,73 @@ def run_main():
+        if os.path.isfile(abi_args.report_dir):
+            print("Error: {} is not a directory".format(abi_args.report_dir))
+            parser.exit()
+        old_version = SimpleNamespace(
```
I'd have expected this to be a `namedtuple`. What's the advantage of using `SimpleNamespace`?

dgreen-arm replied:
As `namedtuples` are immutable, I had thought this would be a problem when adding the `abi_dumps` and `modules`, so used `SimpleNamespace` instead. I was mistaken though, while they're immutable, you can still modify the fields which are dictionaries, which is all I need. I can change them to be `namedtuples` if that would be better.

gilles-peskine-arm replied:
Ok, thanks. Since the structure doesn't change, I think `namedtuple` would be better, but that's very minor.
